compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ClasspathChangesComputer.kt

```diff
@@ -78,7 +78,11 @@ object ClasspathChangesComputer {
         }
 
         return metrics.measure(BuildTime.COMPUTE_IMPACTED_SET) {
-            computeImpactedSet(classChanges, previousClassSnapshots)
+            // classChanges may contain added symbols (they can also impact recompilation -- see examples in JavaClassChangesComputer), so
+            // we need to consider all classes on both the current and previous classpath. It's okay for the combined list to contain
+            // overlapping ClassIds.
+            val allClasses = (currentClassSnapshots.asSequence() + changedPreviousClasses.asSequence()).asIterable()
```

Comment:
Previously only `previousClassSnapshots` was used, and I assume that we could hit some bugs. Can you please add a regression test for that case?

Reply:
Hm, I couldn't come up with a regression test for this. I can't think of an added class that can impact an unchanged class on the classpath (`.class` file, not source file). So it seems that this wasn't a bug before.

I'm leaning towards keeping `allClasses = current + previous classes` just to be safe in case impact analysis involves more than just class hierarchies in the future.  It's also nicer as `classChanges` in the next line will be guaranteed to be a subset `allClasses`.

(Having `allClasses = previous classes only` might be slightly better for performance though.)

WDYT?

Reply:
This comment suggests otherwise, so I'd suggest removing it, or clarifying the actual state (what you have in the comment).

Scenario in which you may have an added class (A) that impacts unchanged class (B) implies that the B referenced non-existent type A in its bytecode (e.g as supertype). However, this would almost certainly cause the compiler to reject loading B. Another option is to have A defined in sources, but in such case within-compilation IC should take care of everything.

Either way, keeping this as-is sounds good, especially because the list of added classes should typically be small.

Reply:
Yes, I've kept the code and added some comments, hopefully it's clearer.